### Method 1: 

Installing the bitcoin core software from the command line using the bitcoin's ppa repository

In [105]:
### Package Installation

# Installing the BitCoin Core Software
!pkexec apt-add-repository ppa:bitcoin/bitcoin -y
!pkexec apt-get update
!pkexec apt-get install bitcoind -y
    

Repository: 'deb https://ppa.launchpadcontent.net/bitcoin/bitcoin/ubuntu/ jammy main'
Description:
NOT MAINTAINED. The OS-library linking packages here had a series of issues.

PLEASE DOWNLOAD DIRECTLY FROM bitcoincore.org (and verify the signatures of said files).

IF YOU WANT AUTO-UPDATES, please see the officially-maintained snap package -
https://github.com/bitcoin-core/packaging/tree/master/snap
More info: https://launchpad.net/~bitcoin/+archive/ubuntu/bitcoin
Adding repository.
Found existing deb entry in /etc/apt/sources.list.d/bitcoin-ubuntu-bitcoin-jammy.list
Adding deb entry to /etc/apt/sources.list.d/bitcoin-ubuntu-bitcoin-jammy.list
Found existing deb-src entry in /etc/apt/sources.list.d/bitcoin-ubuntu-bitcoin-jammy.list
Adding disabled deb-src entry to /etc/apt/sources.list.d/bitcoin-ubuntu-bitcoin-jammy.list
Adding key to /etc/apt/trusted.gpg.d/bitcoin-ubuntu-bitcoin.gpg with fingerprint C70EF1F0305A1ADB9986DBD8D46F45428842CE5E
Hit:1 https://download.docker.com/linux/ubun

Hit:2 https://dl.google.com/linux/chrome/deb stable InRelease                  
Ign:3 https://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.6 InRelease     
Ign:4 https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/6.0 InRelease      
Hit:5 http://apt.pop-os.org/proprietary jammy InRelease                        
Ign:6 https://ppa.launchpadcontent.net/bitcoin/bitcoin/ubuntu jammy InRelease  
Hit:7 https://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.6 Release       
Hit:8 http://apt.pop-os.org/release jammy InRelease                            
Hit:10 http://apt.pop-os.org/ubuntu jammy InRelease                            
Hit:12 https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/6.0 Release       
Get:14 http://apt.pop-os.org/ubuntu jammy-security InRelease [110 kB]          
Err:15 https://ppa.launchpadcontent.net/bitcoin/bitcoin/ubuntu jammy Release   
  404  Not Found [IP: 2620:2d:4000:1::3e 443]
Hit:16 https://europe-west6-apt.pkg.dev/projects/radicle-services radicle-

Method 1 failed because there's no corresponding package(bitcoind). Therefore, manually installing the bitcoin core on my Ubuntu 21.04 following the guide: https://www.cyberithub.com/how-to-install-and-setup-bitcoin-core-on-ubuntu-20-04-lts/ (Method 2)

In [101]:
# Checking if bitcoind is installed or not
!bitcoind --version

Bitcoin Core version v0.20.1
Copyright (C) 2009-2020 The Bitcoin Core developers

Please contribute if you find Bitcoin Core useful. Visit
<https://bitcoincore.org/> for further information about the software.
The source code is available from <https://github.com/bitcoin/bitcoin>.

This is experimental software.
Distributed under the MIT software license, see the accompanying file COPYING
or <https://opensource.org/licenses/MIT>



In [142]:
# Starting the Bitcoin Daemon
!bitcoind -regtest -daemon

Bitcoin Core starting


The Regression Test Network (regtest) is a parallel blockchain to the Bitcoin blockchain. It works almost identically to the Bitcoin network, but it is used for private development. Regtest is most similar to testnet in that the coins on the regtest network are not monetized.

### Wallet Creation

In [177]:
# Creating a wallet named "thapa_sabin" using bitcoin-cli
!bitcoin-cli -regtest createwallet thapa_sabin false

{
  "name": "thapa_sabin",
  "warning": ""
}


In [178]:
#Generating an address to mine bitcoins in the following sections
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getnewaddress

bcrt1q9zdaq2gsm2gjlweaezy4hayjyct2g6kchsvt7d


In [179]:
# Checking the current balance in the wallet: "thapa_sabin"
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getbalance

0.00000000


Since we have just created the wallet and not mined any block yet, the balance is 0.
To get bitocin balance, mining should be done as follows:

### Mining

In [187]:
#Mining 100 blocks to get coins
!bitcoin-cli -regtest -rpcwallet=thapa_sabin generatetoaddress 100 bcrt1q9zdaq2gsm2gjlweaezy4hayjyct2g6kchsvt7d

[
  "18fde654dbbe357e926c4f27f9e9c01a2aa7b3befac08b1518e4b3d9ecdb337d",
  "2f91045fcbeb6ed680956387df7bfa85549fdf47f0eeb5b00e65757322af3cca",
  "0fb495f26eeffcf63fd75922851ad6fc3bcc0e103399a2280d1bdc4314cf1dda",
  "41647151be3d912e7c5a5f723fa21bd49262dc1868b22d1efab71407f38e419f",
  "3a46a34ff61b826f52bbbc5ecf6e65f8ac1912b3c37fa7d2c24b01cf0bf60202",
  "4a41845afca8f1f0c84832a9d5774f5d654eb52190d954f90c94e2649ffe0254",
  "33e3171e6154c334abd264b3818e99a2ee49d19b8a03d869cc29e644149edb6d",
  "2ab622ed9e60d34b987a1d6f68dece542b26753ae2ce11591a36008398494fb7",
  "5677dc7a015d0fe5391eb1244d56a281b8d85cc9024617ac0a544738e9ed050c",
  "763d7294a50a3c7c87c6670235ea181585b6c5afa435279665c6dd705c6a24ac",
  "22d7c8cc94ca0d155387de10dc53218d8454daac087e9ea07cfa887528734cde",
  "39eebb503d508ff9e420612224a5d1f24d2c132e130cf321602b37d698537444",
  "386af02bb10daa9f944c7a42c5a4d938ea468e8f0de6cf3696db68d5a4d6e075",
  "03d07639e81842fd2fa90df695d70a22af8ba400d48adba392d873e261ffdf84",
  "1a5d40a8e9d89b5

### Balance checking and creating a transaction

In [188]:
# Checking the current balance again after mining
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getbalance

312.50005846


In [192]:
# Get block count
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getblockcount

844


We've successfully mined 100 blocks and now the wallet balance has been increased to 312.5 from 0. Now that we have some balance, we can send the balance to some other address. Firstly, generating an address to send the coins to and then sending the coins to the address in the following section.

In [172]:
# Generate another address to send bitcoin to
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getnewaddress

bcrt1qsdcf5tu63kp735xat8x9jyuhmsqz03ux7r7ggf


In [193]:
# Set the transaction fee per kB for this wallet. 
# Overrides the global -paytxfee command line parameter.
!bitcoin-cli -regtest -rpcwallet=thapa_sabin settxfee 0.00001

true


### Transaction

In [195]:
# Creating a transaction by running the sendtoaddress command
!bitcoin-cli -regtest -rpcwallet=thapa_sabin sendtoaddress "bcrt1qsdcf5tu63kp735xat8x9jyuhmsqz03ux7r7ggf" 100


1f2fe8313cb72303c184add38a3820e2d77791ac503ec3677007a2d7eaa96b5f


In [196]:
# Checking the balance after the transaction
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getbalance

212.50003537


Note that the balance has decreased from 312.5 to 212.5 after sending 100 BTC. It means that the transaction has been successful. 

In [197]:
# Again Mining 10 block to get coins
!bitcoin-cli -regtest -rpcwallet=thapa_sabin generatetoaddress 10 bcrt1qam2unem0frpz4ed9fe8v73wnm9tzsnth2ze44a

[
  "35bdd0775204846c3760d3e1d85e9e48d33da8850577a59471cc7e4f8db991ea",
  "172e2e4d61dd7d2c8440865083b222d452016c080004bcf3c025bad6b0da4e29",
  "25c89301d6a3cd4a0359c7671f1ecbf7a3bd9297e3e1109c786a89141b6e2255",
  "12a9871959185d252b6186baf5b522ea55ed77ec7af1f58a95df41a407a95f9f",
  "4860af2843ec7745e15c431f1a0b529b9a8230fdacee50adf28e77b33670e791",
  "05b2155fb508d8b77e0eb0b3fdcde965f1c0b5b63756b8ce89f525c2d971aad9",
  "47fc75aab6c2bff7df646c2deb1ab11e74890458ca565c42ca53c11ae8deea6b",
  "48cfdd0a1050025ace237b3b1e9e6c2a7a1024ca38493d612d71efcd381c6d5c",
  "60c9faca691526ee714ffe97d0340f2fc183b28d0c3e07bbcdd89059dc391935",
  "781a783c3b858c0aa953910cf9365fe772f745764f3c3a975ed03e82bf479e22"
]


In [198]:
# Get block count
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getblockcount

854


In [199]:
# Checking the balance after mining
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getbalance

235.93753537


The block count has been increased to 854 from 844 after mining 10 blocks and the balance has also increased

In [200]:
# Checking the wallet info
!bitcoin-cli -regtest -rpcwallet=thapa_sabin getwalletinfo

{
  "walletname": "thapa_sabin",
  "walletversion": 169900,
  "balance": 235.93753537,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 140.62500000,
  "txcount": 201,
  "keypoololdest": 1676710385,
  "keypoolsize": 1000,
  "hdseedid": "4f3d7f9afcb3771647a7d0d81f071d3e853d3760",
  "keypoolsize_hd_internal": 999,
  "paytxfee": 0.00001000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false
}


## Summary

Using bitcoind's regtest mode, it was possible to create a private Bitcoin cluster with an arbitrary number of nodes. It was very useful for testing as it gave the ability to mine blocks by hand and see the transaction results.
Firstly, bitcoin core was downloaded and unpacked into the system. Then, credentials were added to bitcoin.conf file(created) inside the .bitcoin directory.
Then, we were able to make transactions using the bitcoin-cli daemon.